core/services/feeds/connection_manager.go

```diff
@@ -0,0 +1,160 @@
+package feeds
+
+import (
+	"context"
+	"crypto/ed25519"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/gracefulpanic"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	pb "github.com/smartcontractkit/chainlink/core/services/feeds/proto"
+	"github.com/smartcontractkit/wsrpc"
+)
+
+//go:generate mockery --name ConnectionsManager --output ./mocks/ --case=underscore
+
+type ConnectionsManager interface {
+	Connect(opts ConnectOpts)
+	Disconnect(id int64) error
+	Close()
+	GetClient(id int64) (pb.FeedsManagerClient, error)
+}
+
+// connectionsManager manages the rpc connections to Feeds Manager services
+type connectionsManager struct {
+	mu       sync.Mutex
+	wgClosed sync.WaitGroup
+
+	connections map[int64]*connection
+}
+
+type connection struct {
+	// ctx allows us to cancel any connections which are currently blocking
+	// while waiting to establish a connection to FMS.
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	client pb.FeedsManagerClient
+}
+
+func newConnectionsManager() *connectionsManager {
+	return &connectionsManager{
+		mu:          sync.Mutex{},
+		connections: map[int64]*connection{},
+	}
+}
+
+// ConnectOpts defines the required options to connect to an FMS server
+type ConnectOpts struct {
+	FeedsManagerID int64
+
+	// URI is the URI of the feeds manager
+	URI string
+
+	// Privkey defines the local CSA private key
+	Privkey []byte
+
+	// Pubkey defines the Feeds Manager Service's public key
+	Pubkey []byte
+
+	// Handlers defines the wsrpc Handlers
+	Handlers pb.NodeServiceServer
+
+	// OnConnect defines a callback for when the dial succeeds
+	OnConnect func(pb.FeedsManagerClient)
+}
+
+// Connects to a feeds manager
+func (mgr *connectionsManager) Connect(opts ConnectOpts) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	mgr.wgClosed.Add(1)
+
+	mgr.mu.Lock()
+	mgr.connections[opts.FeedsManagerID] = conn
+	mgr.mu.Unlock()
+
+	go gracefulpanic.WrapRecover(func() {
```

Comment:
which part inside here is blocking? wouldn't we just want to do this synchronously and return an error indicating success/failure on setting up the connection

Reply:
There are two blocks here. The first is the `clientConn` dial which will block initially until the connection is established.
The second is once the connection has been established we use the connection to set up the request handlers (wsrpc is bidirectional since a client can both send and receive requests) and then block on the context cancel.